### PR TITLE
NRM-432: fix magic link ampersand bug

### PR DESCRIPTION
**Pull request by Rhodine-orleans-lindsay**
## What?
- Fix bug where an additional ampersand being added to the end of the magic link when clicked causes an engine error - [NRM-342](https://collaboration.homeoffice.gov.uk/jira/browse/NRM-342)
## Why?
-  In the past an `&` was added to the end of the magic link  in the getPersonalisation function in the email-lookup-sender behaviour to resolve a bug where `?hof-cookie-check` was being added to the end of the req.query.token (https://github.com/UKHomeOffice/modern-slavery/commit/8bd6451a3c874ba9326f61d7254ca5a2d06fcfdb)
- Currently, with some clients, due to a second ampersand being generated, an nginx error is triggered when the user clicks the magic link, preventing them from accessing their form.
## How?
- removed the `&` from the link in the getPersonalisation function in email-lookup-sender behaviour
- refactored the token configuration in check-email-token behaviour so that if req.query.token ended with `?hof-cookie-check`, `?hof-cookie-check` would be replaced with empty string.
- added hof logger to functions to improve logging
- refactored behaviours to use async with try/catch blocks for better error handling
## Testing?
- added a unit test to check `?hof-cookie-check` is replaced if present at the end of a token
- refactored relevant unit tests following async refactoring in behaviours
## Check list

- [x] I have reviewed my own pull request for linting issues (e.g. adding new lines)
- [x] I have written tests (if relevant)
- [x] I have created a JIRA number for my branch
- [x] I have created a JIRA number for my commit
- [x] I have followed the chris beams method for my commit https://cbea.ms/git-commit/
here is an [example commit](https://github.com/UKHomeOfficeForms/hof/commit/810959f391187c7c4af6db262bcd143b50093a6e)
- [x] Ensure drone builds are green especially tests
- [x] I will squash the commits before merging

### CODE REVIEW

**Comment by sulthan-ahmed** [comment]
Is there any way to add a super brief explanation of the bug in the commit title? e.g. NRM-432: fix magic link ampersand bug.  It's just a super minor suggestion

File: apps/verify/behaviours/email-lookup-sender.js [hunk 1]
```diff
@@ -8,28 +8,27 @@ const templateId = config.govukNotify.templateUserAuthId;
...
 const firstStep = '/start';
 const tokenGenerator = require('../models/save-token');
+const logger = require('hof/lib/logger')({ env: config.env });
 
 const getPersonalisation = (host, token) => {
   return {
-    // pass in `&` at the end in case there is another
-    // query e.g. ?hof-cookie-check
-    'link': `http://${host + appPath + firstStep}?token=${token}&`,
+    'link': `http://${host + appPath + firstStep}?token=${token}`,
```
Comment (sulthan-ahmed) [inline apps/verify/behaviours/email-lookup-sender.js:15]: out of scope, I think I found an issue here in prod this should be `https`

File: apps/verify/behaviours/email-lookup-sender.js [hunk 2]
```diff
@@ -8,28 +8,27 @@ const templateId = config.govukNotify.templateUserAuthId;
...
     'host': `http://${host}`,
   };
 };
```
Comment (sulthan-ahmed) [inline apps/verify/behaviours/email-lookup-sender.js:18]: maybe we need something like this
```
const isProduction = config.env === 'production';

const getPersonalisation = (host, token) => ({
  'link': `${isProduction ? 'https' : 'http'}://${host}${config.appPath}/start?token=${token}`,
  'host': `${isProduction ? 'https' : 'http'}://${host}`
});
```

File: apps/verify/behaviours/email-lookup-sender.js [hunk 3]
```diff
@@ -44,22 +43,29 @@ module.exports = superclass => class extends superclass {
...
-      const isRecognisedEmail = emailDomainCheck.isValidDomain(emailDomain);
+        if (!isRecognisedEmail) {
+          req.sessionModel.set('recognised-email', false);
+          return next(err);
+        }
 
-      if (!isRecognisedEmail) {
-        req.sessionModel.set('recognised-email', false);
+        const host = req.get('host');
+        const token = tokenGenerator.save(email, organisation);
```
Comment (sulthan-ahmed) [inline apps/verify/behaviours/email-lookup-sender.js:62]:
```suggestion
        const token = await tokenGenerator.save(email, organisation);
```
you'll have to add the keyword `async` above

File: apps/verify/behaviours/email-lookup-sender.js [hunk 4]
```diff
@@ -44,22 +43,29 @@ module.exports = superclass => class extends superclass {
...
+        await sendEmail(email, host, token);
+        return next();
+      } catch (err) {
+        logger.error('error', `Error saving values ${error}`);
```
Comment (sulthan-ahmed) [inline apps/verify/behaviours/email-lookup-sender.js:66]:
```suggestion
        logger.error('error', `Error in the saveValues method ${error}`);
```

**Review by sulthan-ahmed — APPROVED** [review]
Fantastic work, I've only suggested a few minor things.

We need to fix the https thing as a separate PR, I'd say and JIRA ticket

File: apps/verify/behaviours/email-lookup-sender.js (hunk 4 above)
Comment (Rhodine-orleans-lindsay) [reply apps/verify/behaviours/email-lookup-sender.js]: 👍🏾 Have amended the message now

File: apps/verify/behaviours/email-lookup-sender.js (hunk 3 above)
Comment (Rhodine-orleans-lindsay) [reply apps/verify/behaviours/email-lookup-sender.js]: 👍🏾 Have amended this now

File: apps/verify/behaviours/email-lookup-sender.js [hunk 5]
```diff
@@ -8,28 +8,27 @@ const templateId = config.govukNotify.templateUserAuthId;
...
-    // eslint-disable-next-line no-console
-    .then(console.log(`email sent to ${email}`))
-    .catch(error => {
-      // eslint-disable-next-line no-console
-      console.error(error);
-    });
+    logger.log('info', `email sent to ${email}`);
+  }
+  catch (error) {
+    logger.error(`Error sending email: ${error}`);
```
[the remark this answers is not in the document]
Comment (Rhodine-orleans-lindsay) [reply apps/verify/behaviours/email-lookup-sender.js:30]: 👍🏾 have added this in now

File: apps/verify/behaviours/email-lookup-sender.js (hunk 2 above)
Comment (Rhodine-orleans-lindsay) [reply apps/verify/behaviours/email-lookup-sender.js:18]: 👍🏾 You're right, I think ASC and ACQ switch between http and https in a similar way

**Pull request merged**